qiskit/backends/baseprovider.py

```diff
@@ -17,7 +17,7 @@ class BaseProvider(ABC):
     Base class for a backend provider.
     """
     def __init__(self, *args, **kwargs):
-        pass
+        self.name = 'base'
```

Comment:
Can we re-think the need for having a provider `name` as a required instance attribute? It seems the main reason for introducing the parameter is for helping with the changes of the implementation of `register()`, and by design the `Base` classes should ideally be as minimal as possible to avoid side effects and constraints - please see the note in `IBMQProvider`.

Reply:
You could, but then it would make other code more complicated as one would have to check for a provider name, or whether it even exists or not, and what to do in that case.

Reply:
It is preferable to keep the complexity **outside** the core to not increase coupling.

Reply:
What do you suggest doing if there are multiple IBMQ providers to register/unregister?  I think people are fairly comfortable with the idea of a provider name.

Reply:
The solution depends on the specific use-case. For instance:

* If we don't allow several instances of the same provider, simply talking about them is wrong. There is no identity problem and you can always query `type(provider)`.
* If we do allow several instances of the same provider, then they can be stored in a dictionary as @diego-plan9 suggested you: `{ "name": provider_instance }`.

...

So, would you mind to clarify why do you need `names` so we can come up with the proper solution both meeting your needs and without introducing complexity?

Reply:
In that case, use @diego-plan9's name dictionary strategy instead. The [`DefaultQISKitProvider`](https://github.com/QISKit/qiskit-core/blob/master/qiskit/wrapper/defaultqiskitprovider.py#L18) which is defined as a ["meta provider"](https://github.com/QISKit/qiskit-core/blob/master/qiskit/wrapper/defaultqiskitprovider.py#L20) is intended to add extra functionality like the one you're suggesting. This solution would meet your API usage expectation without introducing extra complexity to the core.